docs/en/latest/plugins/opa.md

```diff
@@ -0,0 +1,275 @@
+
+```shell
+curl -XPUT 'http://127.0.0.1:9080/apisix/admin/routes/r1' \
+--header 'X-API-KEY: <api-key>' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+    "uri": "/*",
+    "plugins": {
+        "opa": {
+            "host": "http://127.0.0.1:8181",
+            "policy": "example1"
+        }
+    },
+    "upstream": {
+        "nodes": {
+            "httpbin.org:80": 1
+        },
+        "type": "roundrobin"
+    }
+}'
+```
+
+Try it out.
+
+```shell
+# Successful request
+curl -i -XGET 127.0.0.1:9080/get
+HTTP/1.1 200 OK
+
+# Failed request
+curl -i -XPOST 127.0.0.1:9080/post
+HTTP/1.1 403 FORBIDDEN
+```
+
+### Complex Use Case (custom response)
+
+Next, let's think about some more complex scenarios.
+
+When you need to return a custom error message for an incorrect request, you can implement it this way.
+
+```shell
+curl -XPUT '127.0.0.1:8181/v1/policies/example2' \
+--header 'Content-Type: text/plain' \
+--data-raw 'package example2
+
+import input.request
+
+default allow = false
+
+allow {
+    request.method == "GET"
+}
+
+# custom response body (Accepts a string or an object, the object will respond as JSON format)
+reason = "test" {
+    not allow
+}
+
+# custom response header (The data of the object can be written in this way)
+headers = {
+    "Location": "http://example.com/auth"
+} {
+    not allow
+}
+
+# custom response status code
+status_code = 302 {
+    not allow
+}'
+```
+
+Update the route and set `opa` plugin's `policy` parameter to `example2`. Then, let's try it.
+
+```shell
+# Successful request
+curl -i -XGET 127.0.0.1:9080/get
+HTTP/1.1 200 OK
+
+# Failed request
+curl -i -XPOST 127.0.0.1:9080/post
+HTTP/1.1 302 FOUND
+Location: http://example.com/auth
+
+test
+```
+
+### Complex Use Case (send APISIX data)
+
+Let's think about another scenario, when your decision needs to use some APISIX data, such as `route`, `consumer`, etc., how should we do it?
+
+Create a simple policy `echo`, which will return the data sent by APISIX to the OPA service as is, so we can simply see them.
+
+```shell
+curl -XPUT '127.0.0.1:8181/v1/policies/echo' \
+--header 'Content-Type: text/plain' \
+--data-raw 'package echo
+
+allow = false
+reason = input'
+```
+
+Next, update the config of the route to enable sending route data.
+
+```shell
+curl -XPUT 'http://127.0.0.1:9080/apisix/admin/routes/r1' \
+--header 'X-API-KEY: <api-key>' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+    "uri": "/*",
+    "plugins": {
+        "opa": {
+            "host": "http://127.0.0.1:8181",
+            "policy": "echo",
+            "with_route": true
+        }
+    },
+    "upstream": {
+        "nodes": {
+            "httpbin.org:80": 1
+        },
+        "type": "roundrobin"
+    }
+}'
+```
+
+Try it. As you can see, we output this data with the help of the custom response body function described above, along with the data from the route.
+
+```shell
+curl -XGET 127.0.0.1:9080/get
```

Comment:
```suggestion
$ curl -XGET 127.0.0.1:9080/get
```

Reply:
added

Reply:
it can indicate the starting point of a shell command to prevent copying multiple lines when copying and pasting 🤣